modules/querier/external/client.go

```diff
@@ -0,0 +1,197 @@
+package external
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/cristalhq/hedgedhttp"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/hedgedmetrics"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/oauth2"
+)
+
+var (
+	metricEndpointDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "tempo",
+		Name:      "querier_external_endpoint_duration_seconds",
+		Help:      "The duration of the external endpoints.",
+		Buckets:   prometheus.DefBuckets,
+	}, []string{"endpoint"})
+	metricExternalHedgedRequests = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "tempo",
+			Name:      "querier_external_endpoint_hedged_roundtrips_total",
+			Help:      "Total number of hedged external requests. Registered as a gauge for code sanity. This is a counter.",
+		},
+	)
+)
+
+type Config struct {
+	Backend string
+
+	HTTPConfig     *HTTPConfig
+	CloudRunConfig *CloudRunConfig
+}
+
+type Client struct {
+	httpClient    *http.Client
+	endpoints     []string
+	tokenProvider tokenProvider
+}
+
+type CloudRunConfig struct {
+	Endpoints         []string      `yaml:"external_endpoints"`
+	HedgeRequestsAt   time.Duration `yaml:"external_hedge_requests_at"`
```

Comment:
Sounds good - I moved the request hedging settings up to the parent 👍

I think "endpoints" is still a reasonable name, since we're still invoking Cloud Run over HTTP. So the semantics of "endpoints" here lines up with HTTP. The term "endpoints" is also used in [the Cloud Run docs](https://cloud.google.com/run/):
> Each Cloud Run service gets an out-of-the-box stable HTTPS endpoint

Re adding more settings: there are no additional settings added here, since we're using the existing HTTP interface. But having a dedicated Cloud Run config will come in handy eventually. For example, it'll give us a place to configure features like [invoking Cloud Run over gRPC](https://cloud.google.com/run/docs/triggering/grpc). I think supporting gRPC will require some changes on the serverless endpoint, but taking some benchmarks and implementing it could be a nice followup feature!